analyzers/src/SonarAnalyzer.CFG/ShimLayer/AnalysisContext/SymbolStartAnalysisContext.cs

```diff
@@ -0,0 +1,161 @@
+            nameof(RegisterSyntaxNodeAction), typeof(VB.SyntaxKind));
+
+        // symbolStartAnalysisContextParameter => ((symbolStartAnalysisContextType)receiverParameter)."propertyName"
+        Func<object, TProperty> CreatePropertyAccessor<TProperty>(string propertyName)
+        {
+            var receiverParameter = Parameter(typeof(object));
+            return Lambda<Func<object, TProperty>>(
+                Property(Convert(receiverParameter, symbolStartAnalysisContextType), propertyName),
+                receiverParameter).Compile();
+        }
+
+        // (object receiverParameter, Action<TContext> registerActionParameter)
+        //     => ((symbolStartAnalysisContextType)symbolStartAnalysisContextParameter)."registrationMethodName"<typeArguments>(registerActionParameter)
+        Action<object, Action<TContext>> CreateRegistrationMethod<TContext>(string registrationMethodName, params Type[] typeArguments)
+        {
+            var receiverParameter = Parameter(typeof(object));
+            var registerActionParameter = Parameter(typeof(Action<TContext>));
+            return Lambda<Action<object, Action<TContext>>>(
+                Call(Convert(receiverParameter, symbolStartAnalysisContextType), registrationMethodName, typeArguments, registerActionParameter),
+                receiverParameter,
+                registerActionParameter).Compile();
+        }
+
+        // (object receiverParameter, Action<TContext> registerActionParameter, TParameter additionalParameter)
+        //     => ((symbolStartAnalysisContextType)symbolStartAnalysisContextParameter)."registrationMethodName"<typeArguments>(registerActionParameter, additionalParameter)
+        Action<object, Action<TContext>, TParameter> CreateRegistrationMethodWithAdditionalParameter<TContext, TParameter>(string registrationMethodName, params Type[] typeArguments)
+        {
+            var receiverParameter = Parameter(typeof(object));
+            var registerActionParameter = Parameter(typeof(Action<TContext>));
+            var additionalParameter = Parameter(typeof(TParameter));
+            return Lambda<Action<object, Action<TContext>, TParameter>>(
+                Call(Convert(receiverParameter, symbolStartAnalysisContextType), registrationMethodName, typeArguments, registerActionParameter, additionalParameter),
+                receiverParameter,
+                registerActionParameter,
+                additionalParameter).Compile();
+        }
+    }
+
+    public SymbolStartAnalysisContext(object roslynSymbolStartAnalysisContext) =>
+        RoslynSymbolStartAnalysisContext = roslynSymbolStartAnalysisContext;
+
+    public void RegisterCodeBlockAction(Action<CodeBlockAnalysisContext> action) =>
+        registerCodeBlockAction(RoslynSymbolStartAnalysisContext, action);
+
+    public void RegisterCodeBlockStartAction<TLanguageKindEnum>(Action<CodeBlockStartAnalysisContext<TLanguageKindEnum>> action) where TLanguageKindEnum : struct
```

Comment:
This is likely to cause problems in the "Commons" layer, as the SyntaxKinds are not available there. I wanted the two API (..Wrapper and Sonar..) to stay as close as possible to the original one.

Reply:
`Microsoft.CodeAnalysis.CSharp.SyntaxKind` and `Microsoft.CodeAnalysis..SyntaxKind` and `Microsoft.CodeAnalysis.VisualBasic.SyntaxKind` are available in `SonarAnalyzer.Common`.

But indeed, we would not be able to call the registration at "common" level. We would be forced to define the registration twice, once for C# in `SonarAnalyzer.CSharp` and once for VB.NET in `SonarAnalyzer.VisualBasic`, because only at those level `TSyntaxKind` would be concretized, and would solve with the right overload.
I also understand the point on keeping the shim layer and the wrapping layer 1-to-1.

So I am OK with the current solution. It's just that I find the implementation a bit peculiar, as it seems an overloading mechanism done by hand.